### PR TITLE
Include "end of input" in error message if SyntaxError happens at end of file

### LINKED ISSUE #2151: When SyntaxError occurs at end of input, note that in the error message
For example:
```
SyntaxError: parse error (line 3, end of input)
```

This would provide a useful hint when the input is truncated.

### PR DESCRIPTION
**Pull request by TheBrokenRail**
Fixes #2151

### CODE REVIEW

**Comment by svaarala** [comment]
@TheBrokenRail Looks good, couple of notes:
- Testcases need an update for the expected output where this matters.
- It would be good to add a specific testcase for this, maybe using eval() to drive the test (cover both a SyntaxError before end-of-input, and at end-of-input).
- I'm not sure if this would work, but it might be cleaner to check if current token is DUK_TOK_EOF, rather than comparing the offset?

File: tests/ecmascript/test-dev-syntax-error-line.js [hunk 1]
```diff
@@ -21,8 +21,7 @@ function test(inp) {
         print('never here');
     } catch (e) {
         // Match against current syntax
-        res = /^.*\(line (\d+)\)$/.exec(e.message);
-        print('line', res ? res[1] : 'n/a');
+        print('line', e.lineNumber);
```
Comment (svaarala) [inline tests/ecmascript/test-dev-syntax-error-line.js]: Here the idea of the testcase is to cover the line number in the `.message` property, not the `.lineNumber` property, so the best fix would be to fix the pattern.

Comment (TheBrokenRail) [reply tests/ecmascript/test-dev-syntax-error-line.js]: Fixed

**Comment by TheBrokenRail** [comment]
It now has a test case, and the "end of input" message is displayed more reliably.

**Comment by TheBrokenRail** [comment]
@svaarala Rebased and added to changelog!

**Comment by svaarala** [comment]
Did you try comparing against DUK_TOK_EOF (as current and/or previous token)? It may be more reliable than the offset comparison. (The current comparison also involves a potential wrap if input length is 0, i.e. length - 1 == -1).

**Comment by svaarala** [comment]
I guess that EOF approach won't work really well, e.g. if the input ends on a truncated token like `"truncated string`, no EOF token will be generated before a SyntaxError. So maybe comparing against the input offset is the best way. Does the current pull include "at end of input" for such a truncated string? I.e.
```
// test.js
print("truncated string
```

With the offset approach, could you change:
```
if (thr->compile_ctx->lex.input_length - 1 <= thr->compile_ctx->curr_token.start_offset) {
```
to:
```
if (thr->compile_ctx->curr_token.start_offset + 1 >= thr->compile_ctx->lex.input_length) {
```

The main point is to avoid an underflow when input_length is (start_offset and input_length are unsigned, `size_t`). Also the comparison (IMO) makes more sense in this order, i.e. adjusted offset exceeds input length.

**Comment by TheBrokenRail** [comment]
@svaarala Done!

**Comment by svaarala** [comment]
Looks good, could you squash the fix and I'll then merge :+1:

**Comment by TheBrokenRail** [comment]
@svaarala Squashed!

**Comment by svaarala** [comment]
Thanks!

**Comment by svaarala** [comment]
Hmm, it seems that this approach doesn't indicate "at end of input" for truncated tokens like:

```
eval('"foo bar');
```

I'll see if this can be improved somehow.  Ideally an invalid token or a truncated token at end of input would generate the syntax error hint.

**Comment by svaarala** [comment]
Managed to find a way to detect the end of input more reliably, I'll open a pull later tonight.

**Pull request merged**